core-job/src/main/java/org/apache/kylin/job/execution/AbstractExecutable.java

```diff
@@ -114,48 +147,62 @@ public final ExecuteResult execute(ExecutableContext executableContext) throws E
         logger.info("Executing AbstractExecutable (" + this.getName() + ")");
 
         Preconditions.checkArgument(executableContext instanceof DefaultContext);
-        ExecuteResult result = null;
-
-        try {
-            onExecuteStart(executableContext);
-            Throwable exception;
-            do {
-                if (retry > 0) {
-                    logger.info("Retry " + retry);
-                }
-                exception = null;
-                result = null;
-                try {
-                    result = doWork(executableContext);
-                } catch (Throwable e) {
-                    logger.error("error running Executable: " + this.toString());
-                    exception = e;
-                }
-                retry++;
-            } while (needRetry(result, exception));
 
-            //check exception in result to avoid retry on ChainedExecutable(only need retry on subtask actually)
-            if (exception != null || result.getThrowable() != null) {
-                onExecuteError(exception, executableContext);
-                throw new ExecuteException(exception);
-            }
+        onExecuteStart(executableContext);
 
-            onExecuteFinished(result, executableContext);
-        } catch (Exception e) {
-            if (isMetaDataPersistException(e)) {
-                handleMetaDataPersistException(e);
+        ExecuteResult result = null;
+        Throwable exception;
+        do {
+            if (retry > 0) {
+                logger.info("Retry " + retry);
             }
-            if (e instanceof ExecuteException) {
-                throw e;
-            } else {
-                throw new ExecuteException(e);
+            exception = null;
+            result = null;
+            try {
+                result = doWork(executableContext);
+            } catch (Throwable e) {
+                logger.error("error running Executable: " + this.toString());
+                exception = e;
             }
+            retry++;
+        } while (needRetry(result, exception));
+
+        if (exception != null) {
+            onExecuteError(exception, executableContext);
+            throw new ExecuteException(exception);
         }
+
+        onExecuteFinishedWithRetry(result, executableContext);
         return result;
     }
 
-    protected void handleMetaDataPersistException(Exception e) {
-        // do nothing.
+    protected void handleMetadataPersistException(ExecutableContext context, Throwable exception) {
+        List<String> users = Lists.newArrayList();
+        final String[] adminDls = context.getConfig().getAdminDls();
```

Comment:
Seems that CollectionUtils.addAll() can be used to simplify the code.

Reply:
Good suggestion. I'll refine it